### PR TITLE
Lower MODULO intrinsic

### PR DESCRIPTION
**Pull request by jeanPerier**
Lower MODULO inline as it is easy and allow covering all real/integer kinds already.
Needed in #633 list.

### CODE REVIEW

**Comment by sscalpone** [comment]
Do we want to preserve the modulo operation until we lower to llvm ir?  I am thinking of asking Cameron to implement modulo in llvm ir too.  Fortran modulo is different from Fortran mod; Fortran mod maps to  LLVM IR rem, I think.

File: flang/lib/Lower/IntrinsicCall.cpp [hunk 1]
```diff
@@ -1400,6 +1402,40 @@ mlir::Value IntrinsicLibrary::genMod(mlir::Type resultType,
   return genRuntimeCall("mod", resultType, args);
 }
 
+// MODULO
+mlir::Value IntrinsicLibrary::genModulo(mlir::Type resultType,
```
Comment (schweitzpgi) [inline flang/lib/Lower/IntrinsicCall.cpp:1406]: Do you want to document the algorithm here?

**Review by schweitzpgi — APPROVED** [review]
Looks good.

There is now a FloorFOp in MLIR.

File: flang/lib/Lower/IntrinsicCall.cpp (hunk 1 above)
Comment (jeanPerier) [reply flang/lib/Lower/IntrinsicCall.cpp:1466]: Sure, I added a comment describing what is the algorithm used for MODULO.

**Comment by jeanPerier** [comment]
> Do we want to preserve the modulo operation until we lower to llvm ir?

It is likely to make vectorizers life easier, but I am not sure how important MODULO is. Note that it has no libnvmath implementation, and therefore not vectorized implementation from what I saw (classic flang has `fort_amodulo` and such directly in Fortran runtime). Do you have reasons to have modulo in llvm ir ? If so, I guess we need it in FIR/MLIR too.

> I'm ok making this a TODO, esp if we file a bug against MLIR to add a floored modulo operator, unless they already have one.

I searched but could not find one. I added a TODO in the comment (not a hard one since the implementation is correct). Filing a request sounds reasonable to me if we think it would have benefits to have it.

> There is now a FloorFOp in MLIR.

Indeed, but I am not sure we want to translate MODULO formula directly, it requires divisions and multiplications. The algorithm I used is based on the truncated modulo (MOD/RemFOp) and does not need to do explicit division/multiplication. So FloorFOp is not useful in this specific context.

**Pull request merged**